### PR TITLE
SmallRye Health: terminate request context properly

### LINKED ISSUE #38866: Sporadic error in custom readiness check using `keycloak-admin-client`: `IllegalStateException: Client is closed`
### Describe the bug

Using `quarkus-keycloak-admin-client-reactive` and `quarkus-smallrye-health`, we have a custom `@Readiness` check in an `@ApplicationScoped` bean that get's the request scoped  instance injected. (In reality there is another `@ApplicationScoped` in between, but that shouldn't matter.)
In the respective readiness check method that `Keycloak` instance is used to get the realm and count its users.

This works 99% of the time (and has worked 100% until a few weeks ago), but every now and then the check fails with:
```
java.lang.IllegalStateException: Client is closed
	at org.jboss.resteasy.reactive.client.impl.ClientImpl.abortIfClosed(ClientImpl.java:239)
	at org.jboss.resteasy.reactive.client.impl.WebTargetImpl.abortIfClosed(WebTargetImpl.java:329)
	at org.jboss.resteasy.reactive.client.impl.WebTargetImpl.resolveTemplate(WebTargetImpl.java:108)
	at org.jboss.resteasy.reactive.client.impl.WebTargetImpl.resolveTemplate(WebTargetImpl.java:25)
	at org.keycloak.admin.client.resource.RealmsResource$$QuarkusRestClientInterface.realm(Unknown Source)
	at org.keycloak.admin.client.Keycloak.realm(Keycloak.java:146)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_ClientProxy.realm(Unknown Source)
	at ...
```
or
```
java.lang.IllegalStateException: Client is closed
	at io.vertx.core.http.impl.HttpClientImpl.checkClosed(HttpClientImpl.java:405)
	at io.vertx.core.http.impl.HttpClientImpl.doRequest(HttpClientImpl.java:281)
	at io.vertx.core.http.impl.HttpClientImpl.request(HttpClientImpl.java:184)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$5.lambda$apply$0(ClientSendRequestHandler.java:434)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at org.jboss.resteasy.reactive.client.AsyncResultUni.subscribe(AsyncResultUni.java:31)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.UniOnItemConsume$UniOnItemComsumeProcessor.onItem(UniOnItemConsume.java:43)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:43)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:38)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemConsume.subscribe(UniOnItemConsume.java:30)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:110)
	at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:88)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler.handle(ClientSendRequestHandler.java:94)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler.handle(ClientSendRequestHandler.java:66)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:285)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:275)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:215)
	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:324)
	at org.keycloak.admin.client.resource.UsersResourcecec060e01a4e8719a5229b25492913298e3846fc54.count(Unknown Source)
	at de.someproject.connector.keycloak.client.KeycloakUserApiClient.readinessCheck(KeycloakUserApiClient.java:40)
	at ...
```

As discussed in https://github.com/quarkusio/quarkus/discussions/38750, I changed [this debug call](https://github.com/quarkusio/quarkus/blob/3.7.3/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientImpl.java#L235) to INFO and added the stacktrace which results in:
```
	at org.jboss.resteasy.reactive.client.impl.ClientImpl.close(ClientImpl.java:236)
	at org.keycloak.admin.client.Keycloak.close(Keycloak.java:177)
	at io.quarkus.keycloak.admin.client.common.AutoCloseableDestroyer.destroy(AutoCloseableDestroyer.java:15)
	at io.quarkus.keycloak.admin.client.common.AutoCloseableDestroyer.destroy(AutoCloseableDestroyer.java:9)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_Bean.doDestroy(Unknown Source)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_Bean.destroy(Unknown Source)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_Bean.destroy(Unknown Source)
	at io.quarkus.arc.impl.AbstractInstanceHandle.destroyInternal(AbstractInstanceHandle.java:78)
	at io.quarkus.arc.impl.ContextInstanceHandleImpl.destroy(ContextInstanceHandleImpl.java:26)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_RequestScoped_ContextInstances.removeEach(Unknown Source)
	at io.quarkus.arc.impl.RequestContext.destroy(RequestContext.java:217)
	at io.quarkus.arc.impl.RequestContext.destroy(RequestContext.java:200)
	at io.quarkus.arc.ManagedContext.terminate(ManagedContext.java:50)
	at io.quarkus.smallrye.health.runtime.SmallRyeHealthHandlerBase.handle(SmallRyeHealthHandlerBase.java:37)
	at io.quarkus.smallrye.health.runtime.SmallRyeReadinessHandler.handle(SmallRyeReadinessHandler.java:8)
	at io.quarkus.smallrye.health.runtime.SmallRyeHealthHandlerBase.handle(SmallRyeHealthHandlerBase.java:23)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.vertx.ext.web.impl.RouterImpl.handleContext(RouterImpl.java:250)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:137)
	at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:62)
	at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:40)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:137)
	at io.quarkus.vertx.http.runtime.security.AbstractHttpAuthorizer.doPermissionCheck(AbstractHttpAuthorizer.java:119)
	at io.quarkus.vertx.http.runtime.security.AbstractHttpAuthorizer.checkPermission(AbstractHttpAuthorizer.java:93)
	at io.quarkus.vertx.http.runtime.security.HttpAuthorizer.checkPermission(HttpAuthorizer.java:22)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$1.handle(HttpSecurityRecorder.java:68)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$1.handle(HttpSecurityRecorder.java:60)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:137)
	at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder$13.handle(ResteasyReactiveRecorder.java:339)
	at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder$13.handle(ResteasyReactiveRecorder.java:332)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:137)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler$3$1.onItem(HttpSecurityRecorder.java:301)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler$3$1.onItem(HttpSecurityRecorder.java:291)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:60)
	at io.smallrye.mutiny.operators.uni.builders.DefaultUniEmitter.complete(DefaultUniEmitter.java:37)
	at io.smallrye.mutiny.groups.UniOnNull.lambda$failWith$1(UniOnNull.java:46)
	at io.smallrye.context.impl.wrappers.SlowContextualBiConsumer.accept(SlowContextualBiConsumer.java:21)
	at io.smallrye.mutiny.groups.UniOnItem.lambda$transformToUni$4(UniOnItem.java:177)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:38)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler$3.onItem(HttpSecurityRecorder.java:291)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler$3.onItem(HttpSecurityRecorder.java:276)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.forwardTo(UniMemoizeOp.java:123)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.notifyAwaiters(UniMemoizeOp.java:135)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.onItem(UniMemoizeOp.java:97)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:60)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:38)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:43)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:43)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:38)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.subscribe(UniMemoizeOp.java:59)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler.handle(HttpSecurityRecorder.java:276)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AbstractAuthenticationHandler.handle(HttpSecurityRecorder.java:224)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:137)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:68)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37)
	at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$2.handle(HttpServerCommonHandlers.java:86)
	at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$2.handle(HttpServerCommonHandlers.java:69)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:147)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:123)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:176)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:174)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:124)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
	at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```

Curiously, only in a _few_ of my many test sessions (session meaning running a readiness loop on a cleanly started server), I got this error instead:
```
jakarta.enterprise.context.ContextNotActiveException: RequestScoped context was not active when trying to obtain a bean instance for a client proxy of SYNTHETIC bean [class=org.keycloak.admin.client.Keycloak, id=UWuGrtXQbAUeC9_5MBr5rdaRQ0Y]
	- you can activate the request context for a specific method using the @ActivateRequestContext interceptor binding
	at io.quarkus.arc.impl.ClientProxies.notActive(ClientProxies.java:70)
	at io.quarkus.arc.impl.ClientProxies.getSingleContextDelegate(ClientProxies.java:30)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_ClientProxy.arc$delegate(Unknown Source)
	at org.keycloak.admin.client.Keycloak_UWuGrtXQbAUeC9_5MBr5rdaRQ0Y_Synthetic_ClientProxy.realm(Unknown Source)
	at de.someproject.idp.connector.keycloak.client.KeycloakUserApiClient.api(KeycloakUserApiClient.java:92)
	at de.someproject.idp.connector.keycloak.client.KeycloakUserApiClient.readinessCheck(KeycloakUserApiClient.java:40)
	at de.someproject.common.rest.client.RestClientWrapper$Interceptor.intercept(RestClientWrapper.java:50)
	at de.someproject.idp.connector.KeycloakIdpConnector.executeReadinessCheck(KeycloakIdpConnector.java:321)
	at de.someproject.common.healthcheck.ConnectorReadinessCheck.call(ConnectorReadinessCheck.java:64)
	at ...
```

So I went ahead and tried placing `@ActivateRequestContext` but failed to find the proper spot for this setup.

### Expected behavior

Consistent results with no error jitter.

### Actual behavior

Failing check every now even though Keycloak is alive and kicking.

### How to Reproduce?

no reproducer for now

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

17.0.10

### Quarkus version or git rev

3.7.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

I ran my test sessions in a local docker compose setup but we also see it on our dev envs (also docker compose) and our Nomad envs - so I'd rule out network issues at this point.

See also discussion: https://github.com/quarkusio/quarkus/discussions/38750

### PR DESCRIPTION
**Pull request by Ladicek**
When the SmallRye Health route handler activates request context on its own, it also needs to terminate it. However, that termination needs to happen only after all health checks complete. That's what this commit does.

- Fixes #38866

### CODE REVIEW

**Review by geoand — APPROVED** [review]
LGTM

**Comment by famod** [comment]
Awesome, thanks! I'll give this a spin with my setup.

**Review by xstefank — APPROVED** [review]

File: extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java [hunk 1]
```diff
@@ -48,6 +49,9 @@ private void doHandle(RoutingContext ctx) {
         Context context = Vertx.currentContext();
         getHealth(reporter, ctx).emitOn(MutinyHelper.executor(context))
                 .subscribe().with(health -> {
+                    if (requestContext != null) {
```
Comment (cescoffier) [inline extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: This should probably be in an `onTermination()`, or be duplicated in the failure callback

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Review by mkouba — APPROVED** [review]

File: extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java (hunk 1 above)
Comment (mkouba) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: Or, which is IMO the best solution, in an end handler added to the `RoutingContext`?

**Comment by gsmet** [comment]
@cescoffier @mkouba @Ladicek I let you come up with the best solution. I think we really need to figure this one out today so that it can be part of the release tomorrow.

File: extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java (hunk 1 above)
Comment (xstefank) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: I originally also used the end handler in my testing without issues.

Comment (cescoffier) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: I think you can have only one endHandler, which means that if you have an interceptor adding one, you can have a problem. Note that this is only when you activate the scope in the code.

Comment (Ladicek) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: Done.

Comment (geoand) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: I think what @Ladicek did here make more sense as it makes it clear where the error handling is as opposed to having to look it up somewhere else

Comment (Ladicek) [reply extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java:52]: If you guys think an end handler is a better solution, I'm totally fine with dropping this. Your call, I don't really know :-)

**Review by cescoffier — APPROVED** [review]
(now with the failure callback set)

I think this is the best way to deal with the request scope created within the SmallRyeHealthHandlerBase.

Sure, an endHandler could work, but you would not know who created the request scope. Here, we know who did create it, and because it creates it, it terminates it. 

Also, if the class is used outside of HTTP, it would still work.

**Comment by mkouba** [comment]
> If you guys think an end handler is a better solution, I'm totally fine with dropping this. Your call, I don't really know :-)

@Ladicek keep it this way. It's a robust enough solution ;-). :+1: 

> Sure, an endHandler could work, but you would not know who created the request scope. Here, we know who did create it, and because it creates it, it terminates it.

You would because the end handler would be registered in the `handle()` method like:

```java
public void handle(RoutingContext ctx) {
        ManagedContext requestContext = Arc.container().requestContext();
        if (requestContext.isActive()) {
            doHandle(ctx);
        } else {
            requestContext.activate();
            try {
                doHandle(ctx);
                ctx.addEndHandler(new Handler<AsyncResult<Void>>() {
                    @Override
                    public void handle(AsyncResult<Void> event) {
                        requestContext.terminate();
                    }
                });
            } catch (Exception e) {
                requestContext.terminate();
            }
        }
    }
```

> Also, if the class is used outside of HTTP, it would still work.

I don't think that `SmallRyeHealthHandlerBase` could ever be used outside HTTP as it works with `io.vertx.ext.web.RoutingContext` :shrug: .

**Comment by xstefank** [comment]
I find this more readable:

```
 requestContext.activate();
            try {
                doHandle(ctx);
            } finally {
                ctx.addEndHandler(voidAsyncResult -> requestContext.terminate());
            }
```

**Comment by famod** [comment]
> Awesome, thanks! I'll give this a spin with my setup.

Didn't really follow the recent EndHandler discussion, but both the initial commit and a onTermination change of mine fixed the issue for me! :rocket:

**Comment by gsmet** [comment]
OK, backport label added, I will make sure it ends up in 3.7.4/3.8.0.

**Comment by gsmet** [comment]
Thanks everyone!

**Comment by gsmet** [comment]
@zakkak see the `ImageMetricsITCase.verifyImageMetrics` failure above.

**Comment by zakkak** [comment]
> @zakkak see the `ImageMetricsITCase.verifyImageMetrics` failure above.

Hi @gsmet, see https://github.com/quarkusio/quarkus/issues/38919

**Pull request merged**